cardano-cli/src/Cardano/CLI/Shelley/Run/Key.hs

```diff
@@ -415,86 +421,198 @@ runConvertITNBip32ToStakeKey (ASigningKeyFile (SigningKeyFile sk)) (OutputFile o
+-- signing key.
+data SomeCardanoAddressSigningKey
+  = ACardanoAddrShelleyPaymentSigningKey !(SigningKey PaymentExtendedKey)
+  | ACardanoAddrShelleyStakeSigningKey !(SigningKey StakeExtendedKey)
+  | ACardanoAddrByronSigningKey !(SigningKey ByronKey)
+
+-- | An error that can occur while converting a @cardano-address@ extended
+-- signing key.
+data CardanoAddressSigningKeyConversionError
+  = CardanoAddressSigningKeyBech32DecodeError !Bech32DecodeError
+  -- ^ There was an error in decoding the string as Bech32.
+  | CardanoAddressSigningKeyDeserialisationError !ByteString
+  -- ^ There was an error in converting the @cardano-address@ extended signing
+  -- key.
+  deriving (Show, Eq)
+
+instance Error CardanoAddressSigningKeyConversionError where
+  displayError = Text.unpack . renderCardanoAddressSigningKeyConversionError
+
+-- | Render an error message for a 'CardanoAddressSigningKeyConversionError'.
+renderCardanoAddressSigningKeyConversionError
+  :: CardanoAddressSigningKeyConversionError
+  -> Text
+renderCardanoAddressSigningKeyConversionError err =
+  case err of
+    CardanoAddressSigningKeyBech32DecodeError decErr ->
+      Text.pack (displayError decErr)
+    CardanoAddressSigningKeyDeserialisationError _bs ->
+      "Error deserialising cardano-address signing key."
+
+-- | Decode a Bech32-encoded string.
+decodeBech32
+  :: Text
+  -> Either Bech32DecodeError (Bech32.HumanReadablePart, Bech32.DataPart, ByteString)
+decodeBech32 bech32Str =
+  case Bech32.decodeLenient bech32Str of
+    Left err -> Left (Bech32DecodingError err)
+    Right (hrPart, dataPart) ->
+      case Bech32.dataPartToBytes dataPart of
+        Nothing ->
+          Left $ Bech32DataPartToBytesError (Bech32.dataPartToText dataPart)
+        Just bs -> Right (hrPart, dataPart, bs)
+
+-- | Convert a Ed25519 BIP32 extended signing key (96 bytes) to a @cardano-crypto@
+-- style extended signing key.
+--
+-- Note that both the ITN and @cardano-address@ use this key format.
+convertBip32SigningKey
+  :: ByteString
+  -> Either CardanoAddressSigningKeyConversionError Crypto.XPrv
+convertBip32SigningKey signingKeyBs =
+  case xprvFromBytes signingKeyBs of
+    Just xPrv -> Right xPrv
+    Nothing ->
+      Left $ CardanoAddressSigningKeyDeserialisationError signingKeyBs
+
+-- | Read a file containing a Bech32-encoded Ed25519 BIP32 extended signing
+-- key.
+readBech32Bip32SigningKeyFile
+  :: SigningKeyFile
+  -> IO (Either (FileError CardanoAddressSigningKeyConversionError) Crypto.XPrv)
+readBech32Bip32SigningKeyFile (SigningKeyFile fp) = do
+  eStr <- Exception.try $ readFile fp
+  case eStr of
+    Left e -> pure . Left $ FileIOError fp e
+    Right str ->
+      case decodeBech32 (Text.concat $ Text.words str) of
+        Left err ->
+          pure $ Left $
+            FileError fp (CardanoAddressSigningKeyBech32DecodeError err)
+        Right (_hrPart, _dataPart, bs) ->
+          pure $ first (FileError fp) (convertBip32SigningKey bs)
```

Comment:
I had a go to see what an `ExceptT` version would look like:

```
readBech32Bip32SigningKeyFile
  :: SigningKeyFile
  -> ExceptT (FileError CardanoAddressSigningKeyConversionError) IO Crypto.XPrv
readBech32Bip32SigningKeyFile (SigningKeyFile fp) = do
  str <- readFile fp                                                      & handleIOExceptT (FileIOError fp)
  (_hrPart, _dataPart, bs) <- decodeBech32 (Text.concat $ Text.words str) & except & withExceptT (FileError fp . CardanoAddressSigningKeyBech32DecodeError)
  convertBip32SigningKey bs                                               & except & withExceptT (FileError fp)
```

The LHS of the `&` would be the business logic and the RHS of `&` is all the error type mangling.